### PR TITLE
[1.x] Implement ratelimiting for sources

### PR DESCRIPTION
**Pull request by Flat**
This will allow sources to specify a rate limit to chain. Requests that are over the rate limit will be dropped and an error returned.

### CODE REVIEW

File: common/src/main/kotlin/http/RatelimitInterceptor.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+package tachiyomi.core.http
+
+import okhttp3.Interceptor
+import okhttp3.Response
+import prefs.RateBucketPreferences
+import tachiyomi.core.di.AppScope
+
+class RatelimitInterceptor(
```
Comment (inorichi) [inline common/src/main/kotlin/http/RatelimitInterceptor.kt]: Make this class receive a `PreferenceStore` as parameter, then you can pass the store assigned to each source through `deps.preferences`

File: core/src/main/kotlin/prefs/RateBucketPreferencesProvider.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import android.content.Context
+import tachiyomi.core.prefs.SharedPreferencesStore
+import toothpick.ProvidesSingletonInScope
+import javax.inject.Inject
+import javax.inject.Provider
+import javax.inject.Singleton
+
+@Singleton
+@ProvidesSingletonInScope
+internal class RateBucketPreferencesProvider @Inject constructor(
```
Comment (inorichi) [inline core/src/main/kotlin/prefs/RateBucketPreferencesProvider.kt]: You won't need this class anymore.

File: common/src/main/kotlin/utils/RateBucket.kt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ */
+
+package utils
+
+import java.util.Calendar
+
+class RateBucket(internal var capacity: Int, internal var refillRate: Long) {
+  internal var tokens: Int = capacity
+  internal var refillTime = Calendar.getInstance().timeInMillis
```
Comment (inorichi) [inline common/src/main/kotlin/utils/RateBucket.kt]: Better use `System.currentTimeMillis()`.

File: common/src/main/kotlin/utils/RateBucket.kt [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      false
+    }
+  }
+
+  fun getTokens(): Int {
+    return this.tokens
+  }
+
+  private fun refill() {
+    val now = Calendar.getInstance().timeInMillis
```
Comment (inorichi) [inline common/src/main/kotlin/utils/RateBucket.kt]: `System.currentTimeMillis()`

File: common/src/main/kotlin/utils/RateBucket.kt [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ *  * Copyright (C) 2018 The Tachiyomi Open Source Project
+ *  *
+ *  * This Source Code Form is subject to the terms of the Mozilla Public
+ *  * License, v. 2.0. If a copy of the MPL was not distributed with this
+ *  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ */
+
+package utils
```
Comment (inorichi) [inline common/src/main/kotlin/utils/RateBucket.kt]: `package tachiyomi.core.utils`

File: common/src/main/kotlin/prefs/RateBucketPreferences.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ *  * Copyright (C) 2018 The Tachiyomi Open Source Project
+ *  *
+ *  * This Source Code Form is subject to the terms of the Mozilla Public
+ *  * License, v. 2.0. If a copy of the MPL was not distributed with this
+ *  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ */
+
+package prefs
```
Comment (inorichi) [inline common/src/main/kotlin/prefs/RateBucketPreferences.kt]: `package tachiyomi.core.prefs`. Although I'm not sure this should belong to prefs. Maybe move both RateBucket and this to `http`.

File: common/src/main/kotlin/utils/RateBucket.kt [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+import java.util.Calendar
+
+class RateBucket(internal var capacity: Int, internal var refillRate: Long) {
+  internal var tokens: Int = capacity
+  internal var refillTime = Calendar.getInstance().timeInMillis
+
+  constructor(capacity: Int, refillRate: Long, tokens: Int, refillTime: Long) : this(
```
Comment (inorichi) [inline common/src/main/kotlin/utils/RateBucket.kt]: You can use a single constructor with default parameters instead.

```kt
class RateBucket(
  internal var capacity: Int,
  internal var refillRate: Long,
  internal var tokens: Int = capacity,
  internal var refillTime: Long = System.currentTimeMillis()
)
```

**Comment by inorichi** [comment]
Can you fix the license headers? There are double asterisks.

It should be:

```
/*
 * Copyright (C) 2018 The Tachiyomi Open Source Project
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
```

**Comment by Flat** [comment]
Fixed. Not sure how that happened. Gonna just blame android studio.

**Pull request merged**